### PR TITLE
Feature/2341/improve parser usability

### PR DESCRIPTION
**Pull request by IhsenBouallegue**
# Improve parser usability

Close: #2341

## Description

**Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?

### CODE REVIEW

File: analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package de.maibornwolff.codecharta.tools.interactivecli
+
+import picocli.CommandLine
+
+interface ParserDialogInterface {
+    fun generateParserArgs(args: Array<String>, commandLine: CommandLine): Array<String>
```
Comment (NearW) [inline analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt:6]: I'd recommend to use `List` instead of `Array`

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 2]
```diff
@@ -70,7 +71,25 @@ class Ccsh : Callable<Void?> {
...
+            if (args.isEmpty() && !isParserUnknown(args, commandLine)) {
+                val selectedParser = ParserService.selectParser(commandLine)
+                println("Executing $selectedParser")
+                ParserService.executeSelectedParser(selectedParser)
+            }
+        }
+
+        private fun isParserUnknown(args: Array<String>, commandLine: CommandLine): Boolean {
+            if (args.isNotEmpty()) {
+                val firstArg = args[0]
```
Comment (NearW) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:88]:
```suggestion
                val firstArg = args.first()
```

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            val subCommands = commandLine.subcommands.values
+            val parsersList = mutableListOf<String>()
+
+            for (subCommand in subCommands) {
+                val parserName: String = subCommand.commandName
+                val parserDescriptions = subCommand.commandSpec.usageMessage().description()
+                val parserDescription = parserDescriptions[0]
+                parsersList.add("$parserName - $parserDescription")
+            }
+            return parsersList
```
Comment (NearW) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt:53]:
```suggestion
            return subCommands.map {
                val parserName = subCommand.commandName
                val parserDescriptions = subCommand.commandSpec.usageMessage().description()
                val parserDescription = parserDescriptions[0]
                "$parserName - $parserDescription"           
            }
```

File: analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+    fun isValidFileName(fileName: String): Boolean
+
+    fun getOutputFileName(fullFileName: String): String {
+        return fullFileName.substringAfterLast("/").substringAfterLast("\\").substringBefore(".")
+    }
+
+    fun checkExtension(fileName: String, expectedFileExtension: String): Boolean {
```
Comment (NearW) [inline analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt:14]: rename to `hasMatchingExtension` or `hasExpectedExtension`. This implies that the function returns something AND that this something is a boolean. `checkExtension` sounds like a function, that does not return anything and throws errors in case, something does not match.

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt [hunk 5]
```diff
@@ -116,16 +119,16 @@ class EdgeProjectBuilder(private val project: Project, private val pathSeparator
...
+            filteredEdges: List<Edge>
     ): MutableMap<String, Any> {
         val aggregatedAttributes: MutableMap<String, Any> = mutableMapOf()
 
         listOfAttributes.forEach { key: String ->
             val attributeType = getAttributeTypeByKey(key)
             val filteredAttribute = filteredEdges.filter { edge: Edge -> edge.attributes.containsKey(key) }
             var aggregatedAttributeValue =
-                filteredAttribute.sumBy { edge: Edge -> edge.attributes[key].toString().toFloat().toInt() }
+                    filteredAttribute.sumBy { edge: Edge -> edge.attributes[key].toString().toFloat().toInt() }
```
Comment (NearW) [inline analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt:131]: What on earth is this conversion? :D I know this is old code, but we should work on that.
Is that not working?
```suggestion
                    filteredAttribute.sumBy { edge: Edge -> edge.attributes[key].toInt() }
```

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            val defaultOutputFileName = getOutputFileName(inputFileName)
+            val outputFileName: String = KInquirer.promptInput(
+                message = "What is the name of the output file?",
+                hint = defaultOutputFileName,
+                default = defaultOutputFileName
+            )
+
+            val defaultPathSeparator = "/"
+            val pathSeparator: String = KInquirer.promptInput(
+                message = "What is the path separator?",
```
Comment (NearW) [inline analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt:27]: Can't we detect this through the operating system?

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+class ParserDialog {
+    companion object : ParserDialogInterface {
+        private const val EXTENSION = "cc.json"
+        override fun generateParserArgs(args: Array<String>, commandLine: CommandLine): Array<String> {
+
+            var inputFileName = KInquirer.promptInput(message = "What is the $EXTENSION file that has to be parsed?")
+            if (inputFileName.substringAfter(".") != EXTENSION)
+                inputFileName += ".$EXTENSION"
+            println("File path: $inputFileName")
```
Comment (NearW) [inline analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt:16]: Not sure, if we should use println when using picocli. If I remember correctly, we use a logger to log to fixed logstreams like info or error.

File: analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt (hunk 1 above)
Comment (IhsenBouallegue) [reply analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt]: Fixed in the next commits.
But is there a specific reason? Because I would still have to cast the result back to an array later.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt (hunk 2 above)
Comment (IhsenBouallegue) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:88]: Fixed in the next commits

File: analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt (hunk 4 above)
Comment (IhsenBouallegue) [reply analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt]: I agree! Fixed in the next commits

File: analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt (hunk 1 above)
Comment (NearW) [reply analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt]: https://stackoverflow.com/a/36263748

The main function is derived from java and uses an Array as input. We can't change that. However, we can decide what we use in our code and we should prefer lists there (see link above).
In that case it's fine to simply convert from array to list and list to array when dealing with the main function.

Comment (IhsenBouallegue) [reply analysis/tools/InteractiveCLI/src/main/kotlin/de/maibornwolff/codecharta/tools/interactivecli/ParserDialogInterface.kt]: Aha got you! thanks for the explanation 👍

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt (hunk 7 above)
Comment (IhsenBouallegue) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt:15]: Thank you for the tip. Indeed I found Kotlin Logger being used, and I changed the println to logger.info.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            val selectedParser: String = KInquirer.promptList(
+                message = "Which parser do you want to execute?",
+                choices = getParserNamesWithDescription(commandLine)
+            )
+
+            return extractParserName(selectedParser)
+        }
+
+        fun executeSelectedParser(selectedParser: String) {
+            val args = emptyArray<String>()
```
[the remark this answers is not in the document]
Comment (IhsenBouallegue) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt:20]: Yes, it is being used on line 26. The parsers main method needs to be called with an empty array, so instead of passing an empty array to each, we declare it once. For now, there is only the edgefilter interactive dialog implemented. But the goal is to implement it in all parser and get rid of the "print("Not implemented yet")".

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt (hunk 5 above)
Comment (IhsenBouallegue) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt]: Thanks for the note @NearW 😄
I finally got around to investigating that. Here is what I discovered:
- *toString()* is needed because `edge.attributes[key]` is of type `any`. So `toFloat()` and `toInt()` can't cast it directly.
- *toFloat()* is needed because the result at that point would be "90.0" and to `toInt()` throws an error when trying to cast it. I assume because of the ".".
So casting to a string avoids the problem with `any` and casting to a float avoids the problem with `.`.

However, I fixed the deprecation problem because of `sumBy` and converted to `sumOf`. So that's an improvement at least 👍

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt [hunk 9]
```diff
@@ -60,9 +59,13 @@ class EdgeProjectBuilder(private val project: Project, private val pathSeparator
     private fun insertEdgeAsNode(nodeEdgeName: String) {
         val nodeFilename = nodeEdgeName.split(pathSeparator).reversed().first()
         val nodePath = nodeEdgeName.split(pathSeparator)
-        val nodeParentPath = nodePath.subList(2, max(2, nodePath.size - 1))
-        val node = Node(nodeFilename, NodeType.File)
-        insertNodeInProjectBuilder(node, nodeParentPath)
+        if (nodePath.size <= 1) {
+            println("No parent node found")
```
Comment (ce-bo) [inline analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt:63]: Im not sure if we should display this message, because it is printed at the command line without any information which edge is affected. So maybe the specified message is a bit confusing, isn't it? What do you think?

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt [hunk 10]
```diff
@@ -116,16 +119,16 @@ class EdgeProjectBuilder(private val project: Project, private val pathSeparator
     }
 
     private fun getAggregatedAttributes(
-        listOfAttributes: MutableList<String>,
-        filteredEdges: List<Edge>
+            listOfAttributes: MutableList<String>,
+            filteredEdges: List<Edge>
```
Comment (ce-bo) [inline analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt:123]:
```suggestion
        listOfAttributes: MutableList<String>,
        filteredEdges: List<Edge>
```

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+private val logger = KotlinLogging.logger {}
+
+class ParserDialog {
+    companion object : ParserDialogInterface {
+        private const val EXTENSION = "cc.json"
+        override fun collectParserArgs(): List<String> {
+
+            var inputFileName = KInquirer.promptInput(message = "What is the $EXTENSION file that has to be parsed?")
+            if (inputFileName.substringAfter(".") != EXTENSION)
```
Comment (ce-bo) [inline analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt:16]: I think this will not work for filenames that have multiple dots included, for instance: my.file.cc.json
I suggest to use substringAfterLast instead.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+            when (selectedParser) {
+                "check" -> print("Not implemented yet")
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt:23]: Maybe we should add a new private helper function in this ParserService that just prints the "Not implemented yet" message => private fun printNotSupported(parserName: string) { ... }

And I would prefer a message like "The interactive usage of parser <Foo> is not supported yet. Please specify the full command to run the parser."

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt (hunk 11 above)
Comment (IhsenBouallegue) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt:16]: After last will pick up only the json and not cc.json. Maybe we do a regex match and make sure cc.json is at the end?

File: analysis/tools/ccsh/src/test/kotlin/de/maibornwolff/codecharta/ccsh/parser/ParserServiceTest.kt [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+        Assertions.assertThat(outContent.toString()).contains("Not implemented yet")
+    }
+
+    @Test
+    fun `should execute sourcemonitorimport parser`() {
+        ParserService.executeSelectedParser("sourcemonitorimport")
+
+        Assertions.assertThat(outContent.toString()).contains("Not implemented yet")
+    }
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/test/kotlin/de/maibornwolff/codecharta/ccsh/parser/ParserServiceTest.kt:94]: Idea for future refactoring of the tests: Use a DataProvider to test the execution of all parsers by only writing one test.
Described here: https://www.baeldung.com/parameterized-tests-junit-5#6-method

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt (hunk 9 above)
Comment (IhsenBouallegue) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt:63]: I am not sure either as I didn't add this change. And honestly I don't know why it's there. It has nothing to do with the CLI.

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt (hunk 11 above)
Comment (ce-bo) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt:16]: Ah ok. I think you can do substringAfterLast(".cc.json")

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt (hunk 9 above)
Comment (BridgeAR) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt:63]: This is something @Hall-Ma and @polinasch ran into during the initial development in tests. If I recall correct it's about a faulty cc structure / path unrelated to each other. In that case, the bug would be hidden. It should probably already throw an error before reaching this point by doing some sanity checks. To at least let the user know something is wrong, this has been added. The message itself should be improved and we might even throw an error here (I am not certain how the tool is used, it might be caught upper layers).

File: analysis/build.gradle [hunk 14]
```diff
@@ -80,7 +81,7 @@ subprojects {
         implementation group: 'org.jetbrains.kotlin', name: 'kotlin-stdlib-jdk8', version: kotlin_version
         implementation group: 'org.jetbrains.kotlin', name: 'kotlin-reflect', version: kotlin_version
         implementation group: 'org.jetbrains.kotlinx', name: 'kotlinx-coroutines-core', version: kotlinx_coroutines_version
-
+        implementation 'com.github.kotlin-inquirer:kotlin-inquirer:0.1.0'
```
[the remark this answers is not in the document]
Comment (IhsenBouallegue) [reply analysis/build.gradle]: You are right. I fixed it and moved it to the packages where it was used instead of all of analysis.

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt (hunk 10 above)
Comment (IhsenBouallegue) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt:123]: Fixed in the next commits 👍

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt (hunk 12 above)
Comment (IhsenBouallegue) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt]: I like this approach 👍  I implemented it in this way and slightly adjusted the tests.

File: analysis/tools/ccsh/src/test/kotlin/de/maibornwolff/codecharta/ccsh/parser/ParserServiceTest.kt (hunk 13 above)
Comment (IhsenBouallegue) [reply analysis/tools/ccsh/src/test/kotlin/de/maibornwolff/codecharta/ccsh/parser/ParserServiceTest.kt]: Awesome! The "not implemented yet" tests are there temporarily, but as soon as we implement more interactive dialogs we should look into parameterized tests 😄

File: analysis/tools/ccsh/build.gradle [hunk 15]
```diff
@@ -55,11 +55,14 @@ dependencies {
...
 
     implementation group: 'info.picocli', name: 'picocli', version: picocli_version
+    implementation group: 'io.github.microutils', name: 'kotlin-logging', version: kotlin_logging_version
 
     testImplementation group: 'junit', name: 'junit', version: junit4_version
     testImplementation group: 'org.assertj', name: 'assertj-core', version: assertj_version
+    testImplementation group: 'io.mockk', name: 'mockk', version: mockk_version
 
     testImplementation("org.junit.platform:junit-platform-runner:$junit_platform_version")
+    implementation "org.jetbrains.kotlin:kotlin-script-runtime:1.6.10"
```
[the remark this answers is not in the document]
Comment (IhsenBouallegue) [reply analysis/tools/ccsh/build.gradle]: I am not sure why this dependency was added. When I removed it, everything worked as intended and no errors showed up anywhere. So I took it out completely.

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt (hunk 11 above)
Comment (IhsenBouallegue) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/ParserDialog.kt]: Maybe 🤔 I went with the regex approach as it was clearer to me. Now it will make sure .cc.json is the extension of the file, unrelated to how many `.` the file has 👍

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt (hunk 9 above)
Comment (ce-bo) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt]: I removed the println and log an error instead. Furthermore, I improved the error message a bit.

**Review by ce-bo — COMMENTED** [review]
LGTM! I had some issues when executing the ccsh. I get the error that the ccsh Main class could not be found, so I could not test it locally. Can someone else confirm this?

**Comment by NearW** [comment]
@ce-bo I'm pretty sure, that this can be tested automatically as well. There are no happy-path e2e-tests right now, but it might make sense to add one for each parser. We'd only check for output on stderr and in that case fail the pipeline.

However, this requires a refactoring, since we mostly use println. A kotlin logger should be used instead. You could also document some logging-guidelines and implement best-practices here.

TL;DR
- 2 issues
- add kotlin logging and use dedicated log levels
- add automatic happy-path e2e test for each parser

**Comment by IhsenBouallegue** [comment]
> @ce-bo I'm pretty sure, that this can be tested automatically as well. There are no happy-path e2e-tests right now, but it might make sense to add one for each parser. We'd only check for output on stderr and in that case fail the pipeline.
> 
> However, this requires a refactoring, since we mostly use println. A kotlin logger should be used instead. You could also document some logging-guidelines and implement best-practices here.
> 
> TL;DR
> 
>     * 2 issues
> 
>     * add kotlin logging and use dedicated log levels
> 
>     * add automatic happy-path e2e test for each parser

I totally agree. Proper logging should be used everywhere. When going over the parsers to implement the interactive dialogs, I will try to refactor as much as possible 😄

**Comment by IhsenBouallegue** [comment]
> LGTM! I had some issues when executing the ccsh. I get the error that the ccsh Main class could not be found, so I could not test it locally. Can someone else confirm this?

After building, you will find the ccsh script under bin in the build folder. In my testing, I navigated to it and executed it simply with `./ccsh`.

File: analysis/build.gradle [hunk 16]
```diff
@@ -32,7 +33,7 @@ buildscript {
 
         mockito_core_version = '3.12.4'
         mockito_kotlin_version = '2.2.0'
-        mokk_version = '1.12.3'
+        mockk_version = '1.12.3'
```
Comment (shaman-apprentice) [inline analysis/build.gradle:36]: Is that on purpose?

File: analysis/build.gradle [hunk 17]
```diff
@@ -68,7 +70,7 @@ allprojects {
     apply plugin: "org.jlleitschuh.gradle.ktlint"
 
     jacoco {
-        toolVersion = jacoco_version
+      toolVersion = jacoco_version
```
Comment (shaman-apprentice) [inline analysis/build.gradle:73]: Why have you changed the spaces here?

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt [hunk 18]
```diff
@@ -9,15 +9,14 @@ import de.maibornwolff.codecharta.model.Path
...
-import kotlin.math.max
 
 class EdgeProjectBuilder(private val project: Project, private val pathSeparator: Char) {
     private val logger = KotlinLogging.logger {}
 
     private val projectBuilder = ProjectBuilder(
-        listOf(MutableNode("root", NodeType.Folder)),
-        mutableListOf(),
-        getAttributeTypes()
+            listOf(MutableNode("root", NodeType.Folder)),
```
Comment (shaman-apprentice) [inline analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt:17]: This seems like one identation to much?

**Review by shaman-apprentice — APPROVED** [review]
early approve. See comments

File: analysis/build.gradle (hunk 16 above)
Comment (IhsenBouallegue) [reply analysis/build.gradle:36]: Yes, it is on purpose. It's for the Mock Kotlin dependency https://mockk.io/

File: analysis/build.gradle (hunk 17 above)
Comment (IhsenBouallegue) [reply analysis/build.gradle:73]: That's weird. Will fix it in the next commit

File: analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt (hunk 18 above)
Comment (IhsenBouallegue) [reply analysis/filter/EdgeFilter/src/main/kotlin/de/maibornwolff/codecharta/filter/edgefilter/EdgeProjectBuilder.kt:17]: It seems like the Intellij formatter is a bit confused. Fixed in the next commit 👍
